cli/azd/pkg/tools/github/github.go

```diff
@@ -323,3 +377,184 @@ func isEnvVarTokenSource(message string) bool {
+		return "", err
+	}
+	defer gzFile.Close()
+
+	gzReader, err := gzip.NewReader(gzFile)
+	if err != nil {
+		return "", err
+	}
+	defer gzReader.Close()
+
+	var extractedAt string
+	// tarReader doesn't need to be closed as it is closed by the gz reader
+	tarReader := tar.NewReader(gzReader)
+	for {
+		fileHeader, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			return extractedAt, fmt.Errorf("did not find gh cli within tar file: %w", err)
+		}
+		if fileHeader == nil {
+			continue
+		}
+		if err != nil {
+			return extractedAt, err
+		}
+		// Tha name contains the path, remove it
+		fileNameParts := strings.Split(fileHeader.Name, "/")
+		fileName := fileNameParts[len(fileNameParts)-1]
+		if fileHeader.Typeflag == tar.TypeReg && fileName == "gh" {
+			filePath := filepath.Join(dst, fileName)
+			ghCliFile, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, os.FileMode(fileHeader.Mode))
+			if err != nil {
+				return extractedAt, err
+			}
+			defer ghCliFile.Close()
+			_, err = io.Copy(ghCliFile, tarReader)
+			if err != nil {
+				return extractedAt, err
+			}
+			extractedAt = filePath
+			break
+		}
+	}
+
+	return extractedAt, nil
+}
+
+// extractGhCli gets the Github cli from either a zip or a tar.gz
+func extractGhCli(src, dst string) (string, error) {
+	if strings.Contains(src, ".zip") {
+		return extractFromZip(src, dst)
+	} else if strings.Contains(src, ".tar.gz") {
+		return extractFromTar(src, dst)
+	}
+
+	// exe rights for file
+	// if err := f.Chmod(osutil.PermissionExecutableFile); err != nil {
+	// 	return err
+	// }
+
+	return "nil", fmt.Errorf("Unknown format")
+}
+
+// GetGitHubCliImplementation defines the contract function to acquire the GitHub cli.
+// The `outputPath` is the destination where the github cli is place it.
+type GetGitHubCliImplementation func(
+	ctx context.Context,
+	transporter policy.Transporter,
+	ghVersion semver.Version,
+	extractImplementation ExtractGitHubCliFromFileImplementation,
+	outputPath string) error
+
+// ExtractGitHubCliFromFileImplementation defines how the cli is extracted
+type ExtractGitHubCliFromFileImplementation func(src, dst string) (string, error)
+
+// downloadGh downloads a given version of GitHub cli from the release site.
+func downloadGh(
+	ctx context.Context,
+	transporter policy.Transporter,
+	ghVersion semver.Version,
+	extractImplementation ExtractGitHubCliFromFileImplementation,
+	path string) error {
+
+	// arm and x86 not supported (similar to bicep)
+	var releaseName string
+	switch runtime.GOOS {
+	case "windows":
+		releaseName = "gh_2.22.1_windows_amd64.zip"
+	case "darwin":
+		releaseName = "gh_2.22.1_macOS_amd64.tar.gz"
+	case "linux":
+		releaseName = "gh_2.22.1_linux_amd64.tar.gz"
+	default:
+		return fmt.Errorf("unsupported platform")
+	}
+
+	//https://github.com/cli/cli/releases/download/v2.22.1/gh_2.22.1_linux_arm64.rpm
+	ghReleaseUrl := fmt.Sprintf("https://github.com/cli/cli/releases/download/v%s/%s", ghVersion, releaseName)
```

Comment:
@ellismg @wbreza @danieljurek Have we run afoul of any known GH release artifacts hiccup in the past? I'm just curious about any past learnings here that could inform the right decision.

I wonder if it's more beneficial if we could instead rely on OS-specific package managers and machine-wide installation. One good thing about package manager approach, is that we could easily generalize for any external tooling we depend on that. The other part is that we can handwave servicing requirements. 

But having standalone installation does make us flexible in how we want to force updates.

Reply:
> @ellismg @wbreza @danieljurek Have we run afoul of any known GH release artifacts hiccup in the past? I'm just curious about any past learnings here that could inform the right decision.

The only thing I can recall seeing in the past where hiccups when we tried to resolve "latest" to a specific release.  That involved an API call (via the GitHub REST API) which was unauthenticated and GitHub does somewhat aggressive rate limiting for unauthenticated requests (on the order of 100 an hour, IIRC).

In practice, we hit this limit more often in CI than on local machines (unlikely for a single dev's machine to send a bunch of unauthenticated requests in an hour, much more likely for a CI runner to run 100 jobs that all want to install some GH Artifact in an hour)

Reply:
@weikanglim , the `gh cli` is build with go, see: https://github.com/cli/cli

This means that, when compiled, all dependencies get shipped within the single binary.  There are not shared-libraries, dlls, etc to worry about. It's all self contained. 

The installation manager or UI installers are only to make sure the binary is placed as part of the PATH and, for windows, as a program than can be uninstalled.  But, it is basically a portable application which uses intallers or package managers to make it look like a classic application.

Reply:
@vhvb1989 If their installation logic ever evolves, we would end up in a bad place. Also, we need to replicate their installation logic 1:1 for this to work correctly. I'll leave this at the team's discretion.

Reply:
Also, what happens to all other supported linux distros? I don't think they're officially supported by `azd`, but it will make it more work for us to add support in the future.

https://github.com/cli/cli/blob/v2.23.0/docs/install_linux.md

Reply:
> If their installation logic ever evolves, we would end up in a bad place

the only time when we would want/consider moving to a newer gh-cli version is if they mention security fixes as part of a release. That's usually the trigger to move old versions to the new one.

Reply:
I will wait to hear from users about this.
After an error, users can always add the binary the want in `~/.azd/bin` to get unblock.

I will start working on switching to `api` and in-house `GitHub login` instead of expending more time on enhancing the `GitHub cli` experience